proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1584 @@
+---
+#@schema/nullable
+#@schema/validation not_null=("Cloud credentials are required.", True)
+credential:
+  name:
+  secretContents:
+    cloud:
+```
+
+_(also considered: [`@schema/required`](#schemarequired))._
+
+#### Use Case: Enumeration
+
+Values can be constrained to be one of a finite site (i.e. of an enumeration).
+
+```yaml
+#@data/values-schema
+---
+volumeSnapshotLocation:
+  spec:
+    #@schema/validation one_of=["aws", "azure", "vsphere"]
+    provider: ""
+```
+
+#### Use Case: Union Structures
+
+An author can require that exactly one of a set of sibling node has a value:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@schema/validation one_not_null=["oidc", "ldap"]
+  config:
+    #@schema/nullable
+    oidc:
+      CLIENT_ID: null #! required if oidc enabled
+      CLIENT_SECRET: null #! required if oidc enabled
+      issuer: null #! <OIDC_IDP_URL> is required if oidc enabled
+    #@schema/nullable
+    ldap:
+      host: null #! <LDAP_HOST> is required if ldap enabed
+      bindDN: null
+      bindPW: null
+```
+
+This approach is simpler than (and therefore recommended over) using a discriminator:
+
+```yaml
+#@data/values
+---
+dex:
+  config:
+    type: "oidc"
+    oidc:
+      CLIENT_ID: null #! required if oidc enabled
+      CLIENT_SECRET: null #! required if oidc enabled
+      issuer: null #! <OIDC_IDP_URL> is required if oidc enabled
+    ldap:
+      host: null #! <LDAP_HOST> is required if ldap enabed
+      bindDN: null
+      bindPW: null
+```
+
+#### Use Case: Multiple Nodes
+
+When a validation involves more than just one Node's value:
+
+```python
+def validate_registry():
+ if data.values.persistence.imageChartStorage.type == "filesystem":
+   if data.values.persistence.persistentVolumeClaim.registry.accessMode == "ReadWriteOnce":
+     data.values.registry.replicas == 1 or assert.fail("The registry replicas must be 1 when the image storage is filesystem and the access mode of persistentVolumeClaim is ReadWriteOnce")
+   end
+ end
+end
+```
+(ref: [vmware-tanzu/community-edition/.../harbor/2.2.3/config/values.star](https://github.com/vmware-tanzu/community-edition/blob/cc7eb24b0e13875a06ac8578cd73f83a217ec4d9/addons/packages/harbor/2.2.3/bundle/config/values.star#L45-L53))
+
+The author places the validation at the lowest-common parent Node:
+
+```yaml
+#@ def one_registry_if_pvc_is_filesystem(val):
+#@   if val.persistence.imageChartStorage.type == "filesystem" and
+#@      val.persistence.persistentVolumeClaim.registry.accessMode == "ReadWriteOnce":
+#@      return val.registry.replicas == 1 \
+#@          or (False, "There can be only one registry replica when the image storage is filesystem and the access mode of persistentVolumeClaim is ReadWriteOnce; there are {} replicas.".format(val.registry.replicas))
+#@   end
+#@ end
+
+#@data/values-schema
+#@schema/validation ("There should be exactly one (1) registry replica if Helm Charts are stored on the filesystem.", one_registry_if_pvc_is_filesystem)
```

Comment:
I think there's going to be a common pattern for this "top-node" functions where sub-nodes on the schema will be validated (or take the value from/if) a top level node is defined. An example is what CNR/TAP is defining as TLK (Top Level Keys).

Reply:
Interesting...

We should talk more about what you see, here... cooking up some examples might be useful/insightful.